#!/bin/bash
 
CHR=$1
CHUNK_START=`printf "%.0f" $2`
CHUNK_END=`printf "%.0f" $3`

# directories
ROOT_DIR=../
ONEK_GP_DIR="${ROOT_DIR}"1000genomesForIMPUTE2/1000GP_Phase3/
PHASED_DATA_DIR="${ROOT_DIR}"output/shapeit/oppera.phased/
RESULTS_DIR="${ROOT_DIR}"output/impute2/chr"${CHR}"/
 
# executable (must be IMPUTE version 2.2.0 or later)
IMPUTE2_EXEC=${ROOT_DIR}bin/impute2

#parameters
NE=20000
 
## MODIFY THE FOLLOWING THREE LINES TO ACCOMODATE OTHER PANELS
# reference data files
GENMAP_FILE="${ONEK_GP_DIR}"genetic_map_chr"${CHR}"_combined_b37.txt
HAPS_FILE="${ONEK_GP_DIR}"1000GP_Phase3_chr"${CHR}".hap.gz
LEGEND_FILE="${ONEK_GP_DIR}"1000GP_Phase3_chr"${CHR}".legend.gz
 
## THESE HAPLOTYPES WOULD BE GENERATED BY THE PREVIOUS SCRIPT
## SELECT ONE FROM METHOD-A AND METHOD-B BELOW
# METHOD-A: haplotypes from IMPUTE2 phasing run
#GWAS_HAPS_FILE=${RESULTS_DIR}gwas_data_chr${CHR}.pos${CHUNK_START}-${CHUNK_END}.phasing.impute2_haps

# METHOD-B: haplotypes from SHAPEIT phasing run
GWAS_HAPS_FILE="${PHASED_DATA_DIR}"chr"${CHR}".phased.haps.gz
 
# main output file
OUTPUT_FILE="${RESULTS_DIR}"gwas_data_chr"${CHR}".pos"${CHUNK_START}"-"${CHUNK_END}".impute2

# create output dir for each chromosome
mkdir -p ${ROOT_DIR}output/impute2/chr${CHR}
 
## impute genotypes from GWAS haplotypes
$IMPUTE2_EXEC \
   -use_prephased_g \
   -m $GENMAP_FILE \
   -known_haps_g $GWAS_HAPS_FILE \
   -h $HAPS_FILE \
   -l $LEGEND_FILE \
   -filt_rules_l 'ALL<0.001' \
   -Ne $NE \
   -int $CHUNK_START $CHUNK_END \
   -o_gz \
   -o $OUTPUT_FILE \
   -allow_large_regions \
   -seed 367946
